tests/std/tests/VSO_0000000_allocator_propagation/test.cpp

```diff
@@ -307,7 +457,7 @@ _CONSTEXPR20 bool test_sequence() {
     test_sequence_copy_assign<Sequence, CopyAlloc<int>>(11, 22, 11); // POCCA, non-equal allocators
     test_sequence_copy_assign<Sequence, CopyEqualAlloc<int>>(11, 22, 11); // POCCA, always-equal allocators
 
-    test_sequence_move_ctor<Sequence>();
+    // test_sequence_move_ctor<Sequence>();

```

Comment:
[WIP] I'm temporarily removing this test; I'm not sure what happens to `vector(vector&&)` under `StationaryAlloc` case.

Reply:
<details>
<summary>Problematic</summary>

```cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <forward_list>
#include <functional>
#include <initializer_list>
#include <iterator>
#include <list>
#include <map>
#include <memory>
#include <new>
#include <set>
#include <string>
#include <type_traits>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

template <class Container>
_CONSTEXPR20 void assert_equal(const Container& cont, initializer_list<typename Container::value_type> il) {
    assert(equal(cont.begin(), cont.end(), il.begin(), il.end()));
}

template <class Container>
_CONSTEXPR20 void assert_is_permutation(const Container& cont, initializer_list<typename Container::value_type> il) {
    assert(is_permutation(cont.begin(), cont.end(), il.begin(), il.end()));
}

class test_leak {
    char* ptr;

public:
    test_leak(const test_leak&) = delete;
    test_leak& operator=(const test_leak&) = delete;

    _CONSTEXPR20 test_leak() noexcept /* terminates */ : ptr(allocator<char>{}.allocate(1)) {}
    _CONSTEXPR20 ~test_leak() {
        allocator<char>{}.deallocate(ptr, 1);
    }
};

template <class T>
class nontrivial_pointer : private test_leak {
public:
    T* ptr;
    _CONSTEXPR20 nontrivial_pointer() noexcept : ptr(nullptr) {}
    _CONSTEXPR20 explicit nontrivial_pointer(T* ptr_) noexcept : ptr(ptr_) {}
    _CONSTEXPR20 /*implicit*/ nontrivial_pointer(nullptr_t) noexcept : ptr(nullptr) {}

    _CONSTEXPR20 nontrivial_pointer(const nontrivial_pointer& other) noexcept : ptr(other.ptr) {}
    template <class U, enable_if_t<is_convertible_v<U*, T*>, int> = 0>
    _CONSTEXPR20 nontrivial_pointer(const nontrivial_pointer<U>& other) noexcept : ptr(other.ptr) {}
    _CONSTEXPR20 nontrivial_pointer& operator=(const nontrivial_pointer& other) noexcept {
        ptr = other.ptr;
        return *this;
    }

    _CONSTEXPR20 ~nontrivial_pointer() = default;

    _CONSTEXPR20 explicit operator bool() const noexcept {
        return static_cast<bool>(ptr);
    }
    _CONSTEXPR20 add_lvalue_reference_t<T> operator*() const noexcept {
        return *ptr;
    }
    _CONSTEXPR20 T* operator->() const noexcept {
        return ptr;
    }
    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 add_lvalue_reference_t<T> operator[](I off) const noexcept {
        return ptr[off];
    }

    _CONSTEXPR20 nontrivial_pointer& operator++() noexcept {
        ++ptr;
        return *this;
    }
    _CONSTEXPR20 nontrivial_pointer operator++(int) noexcept {
        return nontrivial_pointer(ptr++);
    }
    _CONSTEXPR20 nontrivial_pointer& operator--() noexcept {
        --ptr;
        return *this;
    }
    _CONSTEXPR20 nontrivial_pointer operator--(int) noexcept {
        return nontrivial_pointer(ptr--);
    }
    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 nontrivial_pointer& operator+=(I diff) noexcept {
        ptr += diff;
        return *this;
    }
    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 nontrivial_pointer& operator-=(I diff) noexcept {
        ptr -= diff;
        return *this;
    }

    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 friend nontrivial_pointer operator+(nontrivial_pointer ptr, I diff) noexcept {
        return ptr += diff;
    }
    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 friend nontrivial_pointer operator+(I diff, nontrivial_pointer ptr) noexcept {
        return ptr += diff;
    }
    _CONSTEXPR20 friend ptrdiff_t operator-(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr - rhs.ptr;
    }
    template <class I, enable_if_t<is_integral_v<I>, int> = 0>
    _CONSTEXPR20 friend nontrivial_pointer operator-(nontrivial_pointer ptr, I diff) noexcept {
        return ptr -= diff;
    }

    _CONSTEXPR20 friend bool operator==(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr == rhs.ptr;
    }
    _CONSTEXPR20 friend bool operator==(nontrivial_pointer ptr, nullptr_t) noexcept {
        return !static_cast<bool>(ptr.ptr);
    }
    _CONSTEXPR20 friend bool operator==(nullptr_t, nontrivial_pointer ptr) noexcept {
        return !static_cast<bool>(ptr.ptr);
    }
    _CONSTEXPR20 friend bool operator!=(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr != rhs.ptr;
    }
    _CONSTEXPR20 friend bool operator!=(nontrivial_pointer ptr, nullptr_t) noexcept {
        return static_cast<bool>(ptr.ptr);
    }
    _CONSTEXPR20 friend bool operator!=(nullptr_t, nontrivial_pointer ptr) noexcept {
        return static_cast<bool>(ptr.ptr);
    }
    _CONSTEXPR20 friend bool operator<(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr < rhs.ptr;
    }
    _CONSTEXPR20 friend bool operator<=(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr <= rhs.ptr;
    }
    _CONSTEXPR20 friend bool operator>(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr > rhs.ptr;
    }
    _CONSTEXPR20 friend bool operator>=(nontrivial_pointer lhs, nontrivial_pointer rhs) noexcept {
        return lhs.ptr >= rhs.ptr;
    }
};

template <class T, bool IsVoid = is_void_v<T>>
struct impl_pointer_to {
    static constexpr nontrivial_pointer<T> pointer_to(T& r) noexcept {
        return nontrivial_pointer<T>(addressof(r));
    }
};

template <class T>
struct impl_pointer_to<T, true> {
    // no pointer_to for void
};

template <class T>
struct std::pointer_traits<nontrivial_pointer<T>> : public impl_pointer_to<T> {
    using pointer = nontrivial_pointer<T>;
    using element_type = T;
    using difference_type = ptrdiff_t;

    template <class U>
    using rebind = nontrivial_pointer<U>;
};

template <class T>
struct std::iterator_traits<nontrivial_pointer<T>> {
    using iterator_category = random_access_iterator_tag;
    using difference_type = ptrdiff_t;
    using value_type = remove_const_t<T>;
    using reference = add_lvalue_reference_t<T>;
    using pointer = nontrivial_pointer<T>;
};

template <class T, class POCCA, class POCMA, class POCS, class EQUAL>
class MyAlloc {
private:
    size_t _id;

    [[nodiscard]] constexpr size_t equal_id() const noexcept {
        return is_always_equal::value ? 10 : _id;
    }

public:
    [[nodiscard]] constexpr size_t id() const noexcept {
        return _id;
    }

    using value_type = T;
    using pointer = nontrivial_pointer<T>;

    using propagate_on_container_copy_assignment = POCCA;
    using propagate_on_container_move_assignment = POCMA;
    using propagate_on_container_swap = POCS;
    using is_always_equal = EQUAL;

    constexpr explicit MyAlloc(const size_t _id_) : _id(_id_) {}

    template <class U>
    constexpr MyAlloc(const MyAlloc<U, POCCA, POCMA, POCS, EQUAL>& other) noexcept : _id(other.id()) {}

    template <class U>
    [[nodiscard]] constexpr bool operator==(const MyAlloc<U, POCCA, POCMA, POCS, EQUAL>& other) const noexcept {
        return equal_id() == other.equal_id();
    }

    template <class U>
    [[nodiscard]] constexpr bool operator!=(const MyAlloc<U, POCCA, POCMA, POCS, EQUAL>& other) const noexcept {
        return equal_id() != other.equal_id();
    }

    [[nodiscard]] constexpr pointer allocate(const size_t numElements) {
        return pointer(allocator<T>{}.allocate(numElements + equal_id()) + equal_id());
    }

    constexpr void deallocate(pointer const first, const size_t numElements) noexcept {
        allocator<T>{}.deallocate(_Unfancy(first - equal_id()), numElements + equal_id());
    }
};

template <class T>
using StationaryAlloc = MyAlloc<T, false_type, false_type, false_type, false_type>;

_CONSTEXPR20 bool problematic() {
    vector<int, StationaryAlloc<int>> src(StationaryAlloc<int>(11));
    vector<int, StationaryAlloc<int>> dst(move(src)); // <<<

    return true;
}

int main() {
#if _HAS_CXX20
    static_assert(problematic());
#endif // _HAS_CXX20
}
```
</details>

Reply:
Min repro; I'm doubting this is a compiler bug:
```cpp
#include <memory>
#include <type_traits>

using namespace std;

class test_leak {
    char* ptr;

public:
    test_leak(const test_leak&) = delete;
    test_leak& operator=(const test_leak&) = delete;

    constexpr test_leak() noexcept /* terminates */ : ptr(allocator<char>{}.allocate(1)) {}
    constexpr ~test_leak() {
        allocator<char>{}.deallocate(ptr, 1);
    }
};

template <class T>
class nontrivial_pointer : private test_leak {
public:
    T* ptr;
    _CONSTEXPR20 nontrivial_pointer() noexcept : ptr(nullptr) {}
    _CONSTEXPR20 explicit nontrivial_pointer(T* ptr_) noexcept : ptr(ptr_) {}
    _CONSTEXPR20 /*implicit*/ nontrivial_pointer(nullptr_t) noexcept : ptr(nullptr) {}

    _CONSTEXPR20 nontrivial_pointer(const nontrivial_pointer& other) noexcept : ptr(other.ptr) {}
    template <class U, enable_if_t<is_convertible_v<U*, T*>, int> = 0>
    _CONSTEXPR20 nontrivial_pointer(const nontrivial_pointer<U>& other) noexcept : ptr(other.ptr) {}
    _CONSTEXPR20 nontrivial_pointer& operator=(const nontrivial_pointer& other) noexcept {
        ptr = other.ptr;
        return *this;
    }

    _CONSTEXPR20 ~nontrivial_pointer() = default;

    //...
};

using pointer = nontrivial_pointer<int>;

struct al {};

struct val {
    constexpr val() :a(), b(), c() {}
    constexpr val(pointer a_, pointer b_, pointer c_) : a(a_), b(b_), c(c_) {}

    pointer a;
    pointer b;
    pointer c;
};

struct vec {
    constexpr vec(const al& a)noexcept :cp(_One_then_variadic_args_t{}, a) {}
#if 0
    // problematic, allocated storage not being deallocated
    constexpr vec(const vec& v)noexcept : cp(_One_then_variadic_args_t{}, v.cp._Get_first(),
        pointer(v.cp._Myval2.a),
        pointer(v.cp._Myval2.b),
        pointer(v.cp._Myval2.c)
    ) {}
#else
    // ok:
    constexpr vec(const vec& v)noexcept : cp(_One_then_variadic_args_t{}, v.cp._Get_first(),
        v.cp._Myval2.a,
        v.cp._Myval2.b,
        v.cp._Myval2.c
    ) {}
#endif
    // ok, though same as the problematic ctor:
    constexpr void ok2() {
        _Compressed_pair<al, val> cp(_One_then_variadic_args_t{}, al{});
        _Compressed_pair<al, val> cp2(_One_then_variadic_args_t{}, cp._Get_first(),
            pointer(cp._Myval2.a),
            pointer(cp._Myval2.b),
            pointer(cp._Myval2.c)
        );
    }

    _Compressed_pair<al, val> cp;
};

// will turn problematic if using the "#if 0" ctor.
// ok2() is always successful.
_CONSTEXPR20 bool test() {
    vec a(al{});
    vec b(a);

    b.ok2();
    return true;
}

int main() {
#if _HAS_CXX20
    static_assert(test());
#endif // _HAS_CXX20
}
```

Reply:
This commit tries to put off leak check to runtime: https://github.com/microsoft/STL/commit/3f63805ae4074c2aa0bcb409550333f9fe259731.

Reply:
Here is another version that restores `test_sequence_move_ctor` by trying to put off leak check to runtime: https://github.com/microsoft/STL/commit/3f63805ae4074c2aa0bcb409550333f9fe259731.

Reply:
And optionally, I think it will be much easier to only fix the initial bug (`operator=(const&)` in this pr (so the test coverage will be non-trivial pointer and constexpr `test_string`, without bothering with `test_leak` etc), and make another pr to deal with `_Take_contents`) I think the main difficulty is really constexpr test coverage for leakage instead of the bugfix...

Reply:
With `#if 0` changed to `#if 1`, MSVC rejects but Clang accepts your test case; this seems like a compiler bug to me.

Reduced to

```
C:\Temp>type meow.cpp
```
```cpp
#include <memory>

struct test_leak {
    char* ptr;

    constexpr test_leak(int, int) : ptr(std::allocator<char>{}.allocate(1)) {}

    test_leak(const test_leak&)            = delete;
    test_leak& operator=(const test_leak&) = delete;

    constexpr ~test_leak() {
        std::allocator<char>{}.deallocate(ptr, 1);
    }
};

struct holder1 {
    constexpr holder1(const test_leak&) {}
};

struct holder2 {
    holder1 h1;

    constexpr holder2() : h1(test_leak(11, 22)) {}
};

constexpr bool test() {
    holder2 h2;
    return true;
}

int main() {
    static_assert(test());
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++20 meow.cpp
meow.cpp
meow.cpp(32): error C2131: expression did not evaluate to a constant
meow.cpp(6): note: failure was caused by allocated storage not being deallocated

C:\Temp>clang-cl /EHsc /nologo /W4 /std:c++20 meow.cpp

C:\Temp>
```

and reported VSO-1888462.